File: src/rules/paths.py
#!/usr/bin/env python3
"""El hogar de las reglas emitidas — la mitad Python del mecanismo.

Existe por una razon concreta y no por simetria: ``paths/declarations.py`` es
el registro de todo hogar que THYROX resuelve por su cuenta, y su recorrido es
Python. Un hogar nuevo cuyo default no pase por ahi es un default SILENCIOSO —
el defecto que ese modulo existe para cerrar.

El tramo de estado NO se escribe aqui. Lo declara ``workbench.paths.state_dir``
(``THYROX_STATE_DIR``, default ``.claude``), que es donde vive la particion que
DEC-01 fija: producto en ``src/``, estado en ``.claude/``. Componer ``.claude``
a mano seria la segunda fuente de verdad de esa particion.

La mitad TypeScript (``src/rules/paths.ts``) declara las mismas dos constantes;
``tests/rules/test_paths_parity.py`` las ata para que no deriven.
"""
from __future__ import annotations

import sys
from pathlib import Path

# Bootstrap canónico (`paths.reach.BOOTSTRAP`): ascenso con detección hasta
# el marcador, NO `parents[N]`. Un offset acierta a UNA profundidad y falla en
# silencio al mover el archivo; el ascenso sobrevive el cambio de anidamiento.
# Es el único punto donde `paths.reach` todavía no se puede importar — de ahí
# en adelante la raíz sale de `reach.thyrox_root()`, no de más aritmética.
_AQUI = Path(__file__).resolve()
_RAIZ = next((p for p in _AQUI.parents
              if (p / "src" / "paths" / "reach.py").is_file()), None)
if _RAIZ is None:
    raise RuntimeError(f"thyrox: no se encontró src/paths/reach.py sobre {_AQUI}")
sys.path.insert(0, str(_RAIZ / "src"))

from paths.reach import env_value, resolve_home  # noqa: E402
from workbench.paths import repo_of, state_dir  # noqa: E402

#: La variable de familia — declara el hogar de TODOS los clones a la vez.
#: Es el ultimo recurso antes del default, no la entrada principal: ver
#: `rules_home_name` para por que una sola clave no basta.
RULES_DIR_VAR = "THYROX_RULES_DIR"

#: El prefijo de la constante POR CLON. Misma forma que
#: `workbench.WORKBENCH_CLONE_PREFIX`, y no por simetria: la razon esta medida
#: en el docstring de `workbench_home_name` con dos episodios silenciosos.
RULES_CLONE_PREFIX = "THYROX_RULES_"

#: El segmento propio de esta familia, dentro del tramo de estado.
RULES_SEGMENT = "rules"


def rules_home_name(repo: str) -> str:
    """La constante por clon: ``db`` -> ``THYROX_RULES_DB``.

    Existe porque **un solo proceso resuelve varios arboles**, y una variable
    global no puede decir dos verdades a la vez. Medido antes de introducirla:
    con solo `RULES_DIR_VAR`, declarar el hogar de `db` le daba a `api`, a
    `docs`, a `server` y a `ui` **el hogar de db** — las cinco filas de
    `declarations.py` imprimian la misma ruta y ninguna avisaba.

    Ese fallo no revienta: las reglas se emiten al arbol equivocado y el clon
    que las esperaba sigue sin ellas. Es la forma que `workbench_home_name` ya
    registro para `THYROX_ENV_FILE`, un nivel mas abajo.

    La composicion se calca de `workbench_home_name` a proposito: dos familias
    con dos gramaticas de nombre obligarian a quien declara a recordar cual es
    cual, que es como se inventa la tercera.
    """
    return f"{RULES_CLONE_PREFIX}{repo.upper().replace('-', '_')}"


def default_rules_dir(root: str | Path) -> Path:
    """El hogar por defecto de un arbol dado, derivado del segmento declarado."""
    root = Path(root)
    return root / state_dir(root) / RULES_SEGMENT


def consumer_rules_dir(root: str | Path) -> Path:
    """El hogar de las reglas de UN CONSUMIDOR, en tres pasos.

    La precedencia va de lo especifico a lo derivado, que es el mismo criterio
    que `reach` declara: una ruta que alguien escribio para ESTE clon no puede
    quedar anulada por una que se escribio para todos::

        THYROX_RULES_<CLONE>  ->  resuelta contra la raiz de ESE clon
        THYROX_RULES_DIR     ->  resuelta contra la raiz de cada clon
        derivado             ->  <raiz>/<tramo de estado>/rules

    Las dos variables se leen CON la raiz del consumidor como punto de partida,
    para que sea SU ``.env`` el que declare su hogar y no el del proveedor. Y
    su valor pasa por ``resolve_home``, que es lo que hace util a la clave de
    FAMILIA: como segmento relativo —``.claude/rules``— dice lo correcto para
    los cinco clones a la vez, porque se compone sobre la raiz de cada uno.
    Sin esa resolucion la clave de familia solo podia llevar una ruta absoluta,
    y una absoluta le da a los cinco el hogar de uno; medido antes de tenerla:
    las cinco filas de ``declarations.py`` imprimian la ruta de ``db``.

    Una ruta ABSOLUTA declarada en la clave de familia sigue colisionando, y es
    correcto que lo haga: quien escribe una absoluta esta nombrando un sitio
    concreto, no un patron.

    El clon se deriva de la ruta (`repo_of`) y no se pide como parametro: un
    llamador que tuviera que nombrarlo podria nombrar otro, y entonces el
    mecanismo resolveria el hogar de un arbol con la clave de otro — que es el
    defecto que `rules_home_name` cierra, reintroducido por la firma.
    """
    root = Path(root)
    repo = repo_of(root)
    if repo is not None:
        specific = env_value(rules_home_name(repo), root)
        if specific:
            return resolve_home(specific, root)
    family = env_value(RULES_DIR_VAR, root)
    if family:
        return resolve_home(family, root)
    return default_rules_dir(root)

File: src/rules/paths.ts
/**
 * El hogar de las reglas emitidas — mismo criterio y misma forma que
 * `skills/paths.ts`, que a su vez calca `agentsDir()`. Se IMPORTA
 * `thyroxRoot`; duplicar su cuerpo sería la segunda fuente de verdad que
 * `calibration-verified-numbers.md` prohíbe.
 *
 * Dos entradas de entorno en el orden que fija `envValue`: la variable del
 * proceso y la del `.env`. Sin ninguna de las dos, el hogar propio de thyrox
 * — para que el mecanismo sea usable sin configurar nada.
 */
import { join } from 'node:path'
import { resolve } from 'node:path'
import { envValue, reach, resolveHome, thyroxRoot } from '../paths/reach.ts'
import { stateDir } from '../workbench/paths.ts'

/**
 * La variable de FAMILIA — declara el hogar de todos los clones a la vez. Es
 * el ultimo recurso antes del default, no la entrada principal: ver
 * `rulesHomeName` para por que una sola clave no basta.
 */
export const RULES_DIR_VAR = 'THYROX_RULES_DIR'

/** El prefijo de la constante POR CLON. Misma forma que la familia workbench. */
export const RULES_CLONE_PREFIX = 'THYROX_RULES_'

/**
 * La constante por clon: `db` -> `THYROX_RULES_DB`.
 *
 * Existe porque un solo proceso resuelve varios arboles, y una variable global
 * no puede decir dos verdades a la vez. Medido antes de introducirla: con solo
 * `RULES_DIR_VAR`, declarar el hogar de `db` le daba a los otros cuatro clones
 * **el hogar de db**, y ninguna fila del registro avisaba.
 */
export function rulesHomeName(repo: string): string {
  return `${RULES_CLONE_PREFIX}${repo.toUpperCase().replace(/-/g, '_')}`
}

/**
 * El nombre corto del clon al que corresponde una raiz, o `null`.
 *
 * Se resuelve contra el mapa que `reach()` ya declara, no recomponiendo el
 * prefijo del clon: ese prefijo es una decision del consumidor y su derivacion
 * vive en `reach`, asi que rehacerla aqui seria su segunda fuente de verdad.
 *
 * Ciega a: una raiz que NO sea uno de los clones declarados — devuelve `null`,
 * y el llamador cae a la clave de familia. Es la conducta correcta: un arbol
 * que el alcance no declara no tiene clave por clon que leer. Y ciega a un
 * subdirectorio de un clon, que la mitad Python SI resuelve (`repo_of`
 * asciende). Esa asimetria es real y esta declarada, no supuesta.
 */
function repoOfRoot(root: string): string | null {
  const target = resolve(root)
  for (const [repo, path] of Object.entries(reach())) {
    if (resolve(path) === target) return repo
  }
  return null
}

/**
 * El segmento propio de esta familia. El tramo de ESTADO no se escribe aqui:
 * lo declara `stateDir()` (`THYROX_STATE_DIR`, default `.claude`), y componerlo
 * a mano seria la segunda fuente de verdad de la particion que DEC-01 fija —
 * producto en `src/`, estado en `.claude/`.
 */
export const RULES_SEGMENT = 'rules'

/** El hogar por defecto de un arbol dado, derivado del segmento declarado. */
export function defaultRulesDir(root: string, start?: string): string {
  return join(root, stateDir(start ?? root), RULES_SEGMENT)
}

/**
 * NO se verifica que el directorio exista: un hogar declarado y ausente es un
 * hecho del consumidor que su llamador tiene que poder ver.
 */
export function rulesDir(start?: string): string {
  const declared = envValue(RULES_DIR_VAR, start)
  if (declared) return declared
  return defaultRulesDir(thyroxRoot(start), start)
}

/**
 * El hogar de las reglas de UN CONSUMIDOR declarado.
 *
 * Existe aparte de `rulesDir()` por un defecto ya registrado en este arbol
 * (#286): un resolutor que cae al hogar del PROVEEDOR cuando se le pasa la
 * raiz de un consumidor compone una ruta dentro de thyrox y la publica como
 * si fuera del clon. Se midio aqui mismo al estrenar el emisor —
 * `--consumer /home/user/kaupamex-api` resolvia a
 * `/home/user/thyrox/.claude/rules`.
 *
 * La variable sigue mandando, pero se lee CON la raiz del consumidor como
 * punto de partida: asi su `.env` es el que declara su hogar. Y su valor pasa
 * por `resolveHome`: como SEGMENTO relativo, la clave de familia dice lo
 * correcto para los cinco clones a la vez, porque se compone sobre la raiz de
 * cada uno. Una absoluta sigue colisionando, y es correcto — quien escribe una
 * absoluta nombra un sitio concreto, no un patron.
 */
export function consumerRulesDir(root: string): string {
  // De lo especifico a lo derivado: una ruta escrita para ESTE clon no puede
  // quedar anulada por una escrita para todos.
  const repo = repoOfRoot(root)
  if (repo) {
    const specific = envValue(rulesHomeName(repo), root)
    if (specific) return resolveHome(specific, root)
  }
  const family = envValue(RULES_DIR_VAR, root)
  if (family) return resolveHome(family, root)
  return defaultRulesDir(root)
}
